parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.variant;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.util.UUID;
+
+/**
+ * This Variant class holds the Variant-encoded value and metadata binary values.
+ */
+public final class Variant {
+  /** The buffer that contains the Variant value. */
+  final ByteBuffer value;
+
+  /** The buffer that contains the Variant metadata. */
+  final ByteBuffer metadata;
+
+  /**
+   * The threshold to switch from linear search to binary search when looking up a field by key in
+   * an object. This is a performance optimization to avoid the overhead of binary search for a
+   * short list.
+   */
+  static final int BINARY_SEARCH_THRESHOLD = 32;
+
+  public Variant(byte[] value, byte[] metadata) {
+    this(value, 0, metadata, 0);
+  }
+
+  Variant(byte[] value, int valuePos, byte[] metadata, int metadataPos) {
+    this(
+        ByteBuffer.wrap(value, valuePos, value.length - valuePos),
+        ByteBuffer.wrap(metadata, metadataPos, metadata.length - metadataPos));
+  }
+
+  Variant(ByteBuffer value, ByteBuffer metadata) {
+    this.value = value.asReadOnlyBuffer();
+    this.value.mark();
+
+    this.metadata = metadata.asReadOnlyBuffer();
+    this.metadata.mark();
+
+    // There is currently only one allowed version.
+    if ((metadata.get(metadata.position()) & VariantUtil.VERSION_MASK) != VariantUtil.VERSION) {
+      throw new UnsupportedOperationException(String.format(
+          "Unsupported variant metadata version: %02X",
```

Comment:
The only allowed version is 1 and only 4 bits are used? Why encode the unsupported version as 2 hex digits?